packages/integration_test/lib/integration_test.dart

```diff
@@ -11,61 +11,117 @@ import 'package:flutter_test/flutter_test.dart';
+    reporterFactory: (engine) => ResultReporter(engine, resultsCompleter),
+  );
+
+  final Map<String, Object> results = await resultsCompleter.future;
+
+  if (reportResultsToNative) {
+    await _reportResultsToNative(binding, results);
+  }
+  return results;
+}
 
-const String _success = 'success';
+String _formatFailureForPlatform(Failure failure) =>
+    '${failure.error} ${failure.details}';
+
+Future<void> _reportResultsToNative(
+    IntegrationTestWidgetsFlutterBinding binding,
+    Map<String, Object> results) async {
+  binding.results = results;
+  print('Test execution completed: ${binding.results}');
+
+  binding._allTestsPassed
+      .complete(!binding.results.values.any((val) => val is Failure));
+
+  try {
+    binding.callbackManager.cleanup();
+    await IntegrationTestWidgetsFlutterBinding._channel.invokeMethod<void>(
+      'allTestsFinished',
+      <String, dynamic>{
+        'results': {
+          for (final result in binding.results.entries)
+            result.key: result.value is Failure
+                ? _formatFailureForPlatform(result.value)
+                : result.value
+        }
+      },
+    );
+  } on MissingPluginException {
+    print('Warning: integration_test test plugin was not detected.');
+  }
+}
 
 /// A subclass of [LiveTestWidgetsFlutterBinding] that reports tests results
 /// on a channel to adapt them to native instrumentation test format.
 class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
     implements IntegrationTestResults {
-  /// Sets up a listener to report that the tests are finished when everything is
-  /// torn down.
+  /// If [run] is not used, sets up a listener to report that the tests are
+  /// finished when everything is torn down.
+  ///
+  /// This functionality is deprecated – clients are expected to use [run] to
+  /// execute their tests instead.
   IntegrationTestWidgetsFlutterBinding() {
-    // TODO(jackson): Report test results as they arrive
-    tearDownAll(() async {
-      try {
-        // For web integration tests we are not using the
-        // `plugins.flutter.io/integration_test`. Mark the tests as complete
-        // before invoking the channel.
-        if (kIsWeb) {
-          if (!_allTestsPassed.isCompleted) {
-            _allTestsPassed.complete(true);
-          }
-        }
-        callbackManager.cleanup();
-        await _channel.invokeMethod<void>(
-          'allTestsFinished',
-          <String, dynamic>{
-            'results': results.map((name, result) {
-              if (result is Failure) {
-                return MapEntry(name, result.details);
-              }
-              return MapEntry(name, result);
-            })
-          },
-        );
-      } on MissingPluginException {
-        print('Warning: integration_test test plugin was not detected.');
-      }
-      if (!_allTestsPassed.isCompleted) _allTestsPassed.complete(true);
-    });
+    if (!_isUsingLegacyReporting) return;
+
```

Comment:
We should print a warning here that there's a new/better way to do this, probably with a link to instructions about how to do it. For now that link could be in the flutter/flutter/wiki area.

Reply:
Good idea, I added more text and a TODO instead, so that we can follow up after we have the executable to mention using the executable or even link to the readme